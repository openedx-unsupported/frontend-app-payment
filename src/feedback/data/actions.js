export const ADD_MESSAGE = 'ADD_MESSAGE';
export const REMOVE_MESSAGE = 'REMOVE_MESSAGE';

export const addMessage = (code, message, data, severity, fieldName = null) => ({
  type: ADD_MESSAGE,
  payload: {
    code,
    message,
    data,
    severity,
    fieldName,
  },
});

export const removeMessage = id => ({
  type: REMOVE_MESSAGE,
  payload: {
    id,
  },
});
