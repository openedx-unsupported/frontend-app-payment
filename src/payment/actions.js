import { utils } from '../common';

const { AsyncActionType } = utils;

export const FETCH_BASKET = new AsyncActionType('PAYMENT', 'FETCH_BASKET');


export const fetchBasket = () => ({
  type: FETCH_BASKET.BASE,
  payload: {},
});

export const fetchBasketBegin = () => ({
  type: FETCH_BASKET.BEGIN,
});

export const fetchBasketSuccess = result => ({
  type: FETCH_BASKET.SUCCESS,
  payload: result,
});

export const fetchBasketFailure = () => ({
  type: FETCH_BASKET.FAILURE,
});
