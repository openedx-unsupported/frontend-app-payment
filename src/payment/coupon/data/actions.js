import { AsyncActionType } from '../../../common/utils';

export const ADD_COUPON = new AsyncActionType('PAYMENT', 'ADD_COUPON');
export const REMOVE_COUPON = new AsyncActionType('PAYMENT', 'REMOVE_COUPON');
export const UPDATE_COUPON_DRAFT = 'PAYMENT__UPDATE_COUPON_DRAFT';

// ADD COUPON

export const addCoupon = code => ({
  type: ADD_COUPON.BASE,
  payload: {
    code,
  },
});

export const addCouponBegin = () => ({
  type: ADD_COUPON.BEGIN,
});

export const addCouponSuccess = (voucherId, code, benefit) => ({
  type: ADD_COUPON.SUCCESS,
  payload: {
    benefit,
    code,
    voucherId,
  },
});

export const addCouponFailure = errorCode => ({
  type: ADD_COUPON.FAILURE,
  payload: {
    errorCode,
  },
});

// REMOVE COUPON

export const removeCoupon = voucherId => ({
  type: REMOVE_COUPON.BASE,
  payload: {
    voucherId,
  },
});

export const removeCouponBegin = () => ({
  type: REMOVE_COUPON.BEGIN,
});

export const removeCouponSuccess = result => ({
  type: REMOVE_COUPON.SUCCESS,
  payload: {
    result,
  },
});

export const removeCouponFailure = errorCode => ({
  type: REMOVE_COUPON.FAILURE,
  payload: {
    errorCode,
  },
});

// UPDATE COUPON DRAFT

export const updateCouponDraft = code => ({
  type: UPDATE_COUPON_DRAFT,
  payload: {
    code,
  },
});
