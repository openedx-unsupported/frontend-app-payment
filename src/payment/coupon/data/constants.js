export const PERCENTAGE_BENEFIT = 'Percentage';
export const ABSOLUTE_BENEFIT = 'Absolute';
