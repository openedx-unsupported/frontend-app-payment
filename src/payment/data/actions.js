import { createRoutine } from 'redux-saga-routines';

// Routines are action + action creator pairs in a series.
// Actions adhere to the flux standard action format.
// Routines by default are in the form of:
//
// Action                |   Action Creator
// -----------------------------------------------
// fetchBasket.TRIGGER   |   fetchBasket()
// fetchBasket.SUCCESS   |   fetchBasket.success()
// fetchBasket.FAILURE   |   fetchBasket.failure()
// fetchBasket.FULFILL   |   fetchBasket.fulfill()
//
// Created with redux-saga-routines
export const fetchCaptureKey = createRoutine('FETCH_CAPTURE_KEY');
export const submitPayment = createRoutine('SUBMIT_PAYMENT');
export const fetchBasket = createRoutine('FETCH_BASKET');
export const addCoupon = createRoutine('ADD_COUPON');
export const removeCoupon = createRoutine('REMOVE_COUPON');
export const updateQuantity = createRoutine('UPDATE_QUANTITY');

// Actions and their action creators
export const BASKET_DATA_RECEIVED = 'BASKET_DATA_RECEIVED';

export const basketDataReceived = basket => ({
  type: BASKET_DATA_RECEIVED,
  payload: basket,
});

export const BASKET_PROCESSING = 'BASKET_PROCESSING';

export const basketProcessing = isProcessing => ({
  type: BASKET_PROCESSING,
  payload: isProcessing,
});

export const CAPTURE_KEY_PROCESSING = 'CAPTURE_KEY_PROCESSING';

export const captureKeyProcessing = isProcessing => ({
  type: CAPTURE_KEY_PROCESSING,
  payload: isProcessing,
});

export const MICROFORM_STATUS = 'MICROFORM_STATUS';

export const microformStatus = status => ({
  type: MICROFORM_STATUS,
  payload: status,
});

export const CAPTURE_KEY_START_TIMEOUT = 'CAPTURE_KEY_START_TIMEOUT';

export const captureKeyStartTimeout = () => ({
  type: CAPTURE_KEY_START_TIMEOUT,
});
export const CAPTURE_KEY_DATA_RECEIVED = 'CAPTURE_KEY_DATA_RECEIVED';

export const captureKeyDataReceived = captureKey => ({
  type: CAPTURE_KEY_DATA_RECEIVED,
  payload: captureKey,
});
