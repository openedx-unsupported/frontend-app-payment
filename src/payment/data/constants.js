export const ORDER_TYPES = {
  BULK_ENROLLMENT: 'Enrollment Code',
  ENTITLEMENT: 'Course Entitlement',
  SEAT: 'Seat',
};

export const CERTIFICATE_TYPES = {
  VERIFIED: 'verified',
  CREDIT: 'credit',
};

/**
 * Payment State for async payment processing and UI Dialog Control.
 *
 * @see POLLING_PAYMENT_STATES
 *
 * @note: This enum is unique to Commerce Coordinator backend.
 */
export const PAYMENT_STATE = (((webserviceEnum = {
  // The enum as the WS Sees it.
  /**
   * Draft (Checkout) Payment
   */
  CHECKOUT: 'checkout',
  /**
   * Payment Complete
   */
  COMPLETED: 'completed',
  /**
   * Server Side Payment Failure
   */
  FAILED: 'failed',
  /**
   * Payment is Pending
   */
  PENDING: 'pending',
  /**
   * Async Processing of payment is ongoing.
   */
  PROCESSING: 'processing',
}) => ({
  ...webserviceEnum,

  // Our Additions

  /**
   * Default according to Redux initial state. (Should be an alias of an official value)
   *
   * @see PAYMENT_STATE.CHECKOUT
   */
  DEFAULT: webserviceEnum.CHECKOUT,

  /**
   * An HTTP Error has occurred between the client and server, this should not be sent over the line.
   *
   * @note **this is custom to the MFE**, thus a Symbol (which JSON usually skips in serialization)
   */
  HTTP_ERROR: Symbol('mfe-only_http_error'),
}))());

/**
 * An array of payment states that we intend to run polling against
 * @type {(string|Symbol)[]} Values from PAYMENT_STATE
 * @see PAYMENT_STATE
 */
export const POLLING_PAYMENT_STATES = [
  PAYMENT_STATE.PROCESSING,
  PAYMENT_STATE.HTTP_ERROR,
];
