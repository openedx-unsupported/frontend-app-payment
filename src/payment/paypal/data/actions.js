import { utils } from '../../../common';

const { AsyncActionType } = utils;

export const SUBMIT_PAYMENT_PAYPAL = new AsyncActionType('PAYMENT', 'SUBMIT_PAYMENT_PAYPAL');

export const submitPaymentPayPal = () => ({
  type: SUBMIT_PAYMENT_PAYPAL.BASE,
});

export const submitPaymentPayPalBegin = () => ({
  type: SUBMIT_PAYMENT_PAYPAL.BEGIN,
});

export const submitPaymentPayPalFailure = () => ({
  type: SUBMIT_PAYMENT_PAYPAL.FAILURE,
});
