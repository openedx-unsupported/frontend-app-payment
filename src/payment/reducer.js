import { FETCH_BASKET } from './actions';

export const initialState = {
  loading: false,
  loadingError: null,
};

const paymentsPage = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_BASKET.BEGIN:
      return {
        ...state,
        loadingError: null,
        loading: true,
      };
    case FETCH_BASKET.SUCCESS:
      return {
        ...state,
        ...action.payload,
        loading: false,
      };
    case FETCH_BASKET.ERROR:
      return {
        ...state,
        loadingError: action.payload,
        loading: false,
      };
    default:
      return state;
  }
};

export default paymentsPage;
