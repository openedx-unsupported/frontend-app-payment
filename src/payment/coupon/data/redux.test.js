import { createStore } from 'redux';

import reducer from './reducers';
import { addCouponBegin, addCouponSuccess, addCouponFailure, removeCouponBegin, removeCouponSuccess, removeCouponFailure, updateCouponDraft, ADD_COUPON, addCoupon, removeCoupon, REMOVE_COUPON } from './actions';
import { PERCENTAGE_BENEFIT } from './constants';

describe('redux tests', () => {
  let store;
  beforeEach(() => {
    store = createStore(reducer);
  });

  it('should return default state in reducer if no action', () => {
    expect(reducer()).toEqual({
      benefit: null,
      code: null,
      voucherId: null,
      error: null,
      loaded: false,
      loading: false,
    });
  });

  it('should match default state', () => {
    expect(store.getState()).toMatchSnapshot();
  });

  it('should begin to add a coupon', () => {
    store.dispatch(addCouponBegin());
    expect(store.getState()).toMatchSnapshot();
  });

  it('should successfully add a coupon', () => {
    store.dispatch(addCouponSuccess(12345, 'DEMO25', { type: PERCENTAGE_BENEFIT, value: 25 }));
    expect(store.getState()).toMatchSnapshot();
  });

  it('should fail to add a coupon', () => {
    store.dispatch(addCouponFailure('oh noes'));
    expect(store.getState()).toMatchSnapshot();
  });

  it('should begin to remove a coupon', () => {
    store.dispatch(addCouponSuccess(12345, 'DEMO25', { type: PERCENTAGE_BENEFIT, value: 25 }));
    store.dispatch(removeCouponBegin());
    expect(store.getState()).toMatchSnapshot();
  });

  it('should successfully remove a coupon', () => {
    store.dispatch(addCouponSuccess(12345, 'DEMO25', { type: PERCENTAGE_BENEFIT, value: 25 }));
    store.dispatch(removeCouponSuccess());
    expect(store.getState()).toMatchSnapshot();
  });

  it('should fail to remove a coupon', () => {
    store.dispatch(addCouponSuccess(12345, 'DEMO25', { type: PERCENTAGE_BENEFIT, value: 25 }));
    store.dispatch(removeCouponFailure('oh noes it didnt'));
    expect(store.getState()).toMatchSnapshot();
  });

  it('should update the code draft', () => {
    store.dispatch(updateCouponDraft('DEM'));
    expect(store.getState()).toMatchSnapshot();
    store.dispatch(updateCouponDraft('DEMO25'));
    expect(store.getState()).toMatchSnapshot();
  });

  /**
   * This action is tested independently because the reducer doesn't directly respond to it.
   * It's for the saga.
   */
  it('should produce an addCoupon action', () => {
    expect(addCoupon('DEMO25')).toEqual({
      type: ADD_COUPON.BASE,
      payload: {
        code: 'DEMO25',
      },
    });
  });

  /**
   * This action is tested independently because the reducer doesn't directly respond to it.
   * It's for the saga.
   */
  it('should produce an removeCoupon action', () => {
    expect(removeCoupon(12345)).toEqual({
      type: REMOVE_COUPON.BASE,
      payload: {
        voucherId: 12345,
      },
    });
  });
});
