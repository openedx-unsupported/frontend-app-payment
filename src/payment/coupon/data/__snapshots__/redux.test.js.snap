// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux tests should begin to add a coupon 1`] = `
Object {
  "benefit": null,
  "code": null,
  "error": null,
  "loaded": false,
  "loading": true,
  "voucherId": null,
}
`;

exports[`redux tests should begin to remove a coupon 1`] = `
Object {
  "benefit": Object {
    "type": "Percentage",
    "value": 25,
  },
  "code": "DEMO25",
  "error": null,
  "loaded": false,
  "loading": true,
  "voucherId": 12345,
}
`;

exports[`redux tests should fail to add a coupon 1`] = `
Object {
  "benefit": null,
  "code": null,
  "error": "oh noes",
  "loaded": false,
  "loading": false,
  "voucherId": null,
}
`;

exports[`redux tests should fail to remove a coupon 1`] = `
Object {
  "benefit": Object {
    "type": "Percentage",
    "value": 25,
  },
  "code": "DEMO25",
  "error": "oh noes it didnt",
  "loaded": true,
  "loading": false,
  "voucherId": 12345,
}
`;

exports[`redux tests should match default state 1`] = `
Object {
  "benefit": null,
  "code": null,
  "error": null,
  "loaded": false,
  "loading": false,
  "voucherId": null,
}
`;

exports[`redux tests should successfully add a coupon 1`] = `
Object {
  "benefit": Object {
    "type": "Percentage",
    "value": 25,
  },
  "code": "DEMO25",
  "error": null,
  "loaded": true,
  "loading": false,
  "voucherId": 12345,
}
`;

exports[`redux tests should successfully remove a coupon 1`] = `
Object {
  "benefit": null,
  "code": null,
  "error": null,
  "loaded": false,
  "loading": false,
  "voucherId": null,
}
`;

exports[`redux tests should update the code draft 1`] = `
Object {
  "benefit": null,
  "code": "DEM",
  "error": null,
  "loaded": false,
  "loading": false,
  "voucherId": null,
}
`;

exports[`redux tests should update the code draft 2`] = `
Object {
  "benefit": null,
  "code": "DEMO25",
  "error": null,
  "loaded": false,
  "loading": false,
  "voucherId": null,
}
`;
